### ISSUE TITLE
Serialization

### ISSUE THREAD

**Issue by einorler** [ongr-io/ElasticsearchBundle#888]
The usage of symfony serializer in 6.0 was not a good idea. There are a couple of issues with it, first of all - the bundle is only compatible with 4.x version of the serializer, which conflicts with SF 3.x that we still claim support for. Another, even more pressing issue: it doesn't work as one would expect it to. It only covers the base case where you only have mapped properties with their respective getters and setters. You basically cannot have any helper methods, because any methods begining with get or is are treated as getters and will break the serializer if there are no corresponding properties. Also, it does not take into account the fact that there can be non-mapped properties and so, serializes everything to ES. This raises the issue of dynamic properties and so on. In addition to that, I don't like the current approach of denying the posibility of singular embeded objects. All these should be addressed in 6.1. I will reintroduce the native object serialization / deserialization with a couple of tweeks.

**Comment by tominventisbe** [comment]
I'm running into the same problem now. I have several helper methods and, even though these are ignored during index creation, when documents are actually indexed, additional fields are created for these helpers, i.e. `getDateAsObject` creates a new field `dateAsObject`, even though no such property exists.

The bundle already knows what properties are part of the index via annotations. Would it be a possibility to simply do this:

```php
public function convertDocumentToArray($document, Serializer $serializer): array
{
    return $serializer->normalize($document, 'array', [
            AbstractNormalizer::ATTRIBUTES => ['propertyWithAnnotation1', 'propertyWithAnnotation2'],
    ]);
}
```

In other words, you _can_ ignore the properties during serialization, but you need to explicitly do this for every document, which is a bit cumbersome, currently.

This leaves the problem that Symfony will try to unserialize to the types you expect. For example, if you expect a `DateTimeImmutable $date` in your constructor and set `$this->date = $date->format('c')`, unserialization will later fail because a string comes back from the index and Symfony Serializer wants to turn that into whatever the constructor expects, which is a `DateTimeImmutable`, which will then fail.

**Comment by toooni** [comment]
I created a PR which solves this issue https://github.com/ongr-io/ElasticsearchBundle/pull/903

**Comment by toooni** [comment]
@saimaz I just saw that https://github.com/ongr-io/ElasticsearchBundle/pull/889 already solved this issue for 6.1 but I am not sure which one is the better solution.

**Comment by saimaz** [comment]
@einorler version looks very good IMO, closing this issue.
